libompd/src/omp-icv.cpp

```diff
@@ -236,6 +237,157 @@ static ompd_rc_t ompd_get_debug(
   return ret;
 }
 
+/* Helper routine for the ompd_get_nthreads routines */
+static ompd_rc_t ompd_get_nthreads_aux(
+    ompd_thread_handle_t *thread_handle,
+    uint32_t *used,
+    uint32_t *current_nesting_level,
+    uint32_t *nproc
+    ) {
+  if (!thread_handle)
+    return ompd_rc_stale_handle;
+  if (!thread_handle->ah)
+    return ompd_rc_stale_handle;
+  ompd_address_space_context_t *context = thread_handle->ah->context;
+  if (!context)
+    return ompd_rc_stale_handle;
+  if (!callbacks)
+    return ompd_rc_error;
+
+  ompd_rc_t ret = TValue(context, "__kmp_nested_nth")
+           .cast("kmp_nested_nthreads_t")
+           .access("used")
+           .castBase(ompd_type_int)
+           .getValue(*used);
+  if (ret != ompd_rc_ok)
+    return ret;
+
+  TValue taskdata =
+      TValue(context, thread_handle->th) /*__kmp_threads[t]->th*/
+          .cast("kmp_base_info_t")
+          .access("th_current_task") /*__kmp_threads[t]->th.th_current_task*/
+          .cast("kmp_taskdata_t", 1);
+
+  ret = taskdata
+          .access("td_team") /*__kmp_threads[t]->th.th_current_task.td_team*/
+          .cast("kmp_team_p", 1)
+          .access("t") /*__kmp_threads[t]->th.th_current_task.td_team->t*/
+          .cast("kmp_base_team_t", 0) /*t*/
+          .access("t_level")          /*t.t_level*/
+          .castBase(ompd_type_int)
+          .getValue(*current_nesting_level);
+  if (ret != ompd_rc_ok)
+    return ret;
+
+  ret = taskdata
+          .cast("kmp_taskdata_t", 1)
+          .access("td_icvs") /*__kmp_threads[t]->th.th_current_task->td_icvs*/
+          .cast("kmp_internal_control_t", 0)
+          .access("nproc") /*__kmp_threads[t]->th.th_current_task->td_icvs.nproc*/
+          .castBase(ompd_type_int)
+          .getValue(*nproc);
+  if (ret != ompd_rc_ok)
+    return ret;
+
+  return ompd_rc_ok;
+}
+
+static ompd_rc_t ompd_get_nthreads(
+    ompd_thread_handle_t *thread_handle, /* IN: handle for the thread */
+    ompd_word_t *nthreads_var_val        /* OUT: nthreads-var (of integer type)
+                                            value */
+    ) {
+  uint32_t used;
+  uint32_t nproc;
+  uint32_t current_nesting_level;
+
+  ompd_rc_t ret;
+  ret = ompd_get_nthreads_aux (thread_handle, &used,
+                               &current_nesting_level, &nproc);
+  if (ret != ompd_rc_ok)
+    return ret;
+
+  /*__kmp_threads[t]->th.th_current_task->td_icvs.nproc*/
+  *nthreads_var_val = nproc;
+  /* If the nthreads-var is a list with more than one element, then the value of
+     this ICV cannot be represented by an integer type. In this case,
+     ompd_rc_incompatible is returned. The tool can check the return value and
+     can choose to invoke ompd_get_icv_string_from_scope() if needed. */
+  if (current_nesting_level < used - 1) {
+    return ompd_rc_incompatible;
```

Comment:
Note for OMPD specification: we might add a new return code: `ompd_rc_incomplete`.
For the currently available return codes, I would not expect meaningful values unless `ompd_rc_ok`.